File: rxflux2-kotlin/src/main/java/com/huyingbao/rxflux2/dispatcher/RxBus.java
package com.huyingbao.rxflux2.dispatcher;

import io.reactivex.Flowable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;

public class RxBus {
    private final FlowableProcessor<Object> mBus;
    private static RxBus sInstance;

    private RxBus() {
        mBus = PublishProcessor.create().toSerialized();
    }

    public synchronized static RxBus getInstance() {
        if (sInstance == null) sInstance = new RxBus();
        return sInstance;
    }

    public FlowableProcessor<Object> get() {
        return mBus;
    }

    public void send(Object o) {
        mBus.onNext(o);
    }

    /**
     * 返回指定类型的带背压的Flowable实例
     *
     * @param tClass
     * @param <T>
     * @return
     */
    public <T> Flowable<T> toFlowable(Class<T> tClass) {
        return mBus.ofType(tClass);
    }

    public void unregisterAll() {
        mBus.onComplete();
    }

    /**
     * 是否已有观察者订阅
     *
     * @return
     */
    public boolean hasSubscribers() {
        return mBus.hasSubscribers();
    }
}

File: rxflux2-kotlin/src/main/java/com/huyingbao/rxflux2/dispatcher/RxBus.kt
package com.huyingbao.rxflux2.dispatcher

import io.reactivex.Flowable
import io.reactivex.processors.FlowableProcessor
import io.reactivex.processors.PublishProcessor

class RxBus private constructor() {
    private val mBus: FlowableProcessor<Any>

    init {
        mBus = PublishProcessor.create<Any>().toSerialized()
    }

    fun get(): FlowableProcessor<Any> {
        return mBus
    }

    fun send(o: Any) {
        mBus.onNext(o)
    }

    /**
     * 返回指定类型的带背压的Flowable实例
     *
     * @param tClass
     * @param <T>
     * @return
    </T> */
    fun <T> toFlowable(tClass: Class<T>): Flowable<T> {
        return mBus.ofType(tClass)
    }

    fun unregisterAll() {
        mBus.onComplete()
    }

    /**
     * 是否已有观察者订阅
     *
     * @return
     */
    fun hasSubscribers(): Boolean {
        return mBus.hasSubscribers()
    }

    companion object {
        private var sInstance: RxBus? = null

        val instance: RxBus
            @Synchronized get() {
                if (sInstance == null) sInstance = RxBus()
                return sInstance
            }
    }
}
